liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,889 @@
+                }
+                long timeout = verifyTimeout * 1000;
+                long endTime = System.currentTimeMillis() + timeout;
+
+                // Wait for the app started message in messages.log
+                String startMessage = serverTask.waitForStringInLog(START_APP_MESSAGE_REGEXP, timeout, messagesLogFile);
+                if (startMessage == null) {
+                    stopServer();
+                    throw new MojoExecutionException(
+                            MessageFormat.format(messages.getString("error.server.start.verify"), verifyTimeout));
+                }
+
+                timeout = endTime - System.currentTimeMillis();
+            } catch (Exception e) {
+                log.debug("Error starting server", e);
+            }
+        }
+
+        @Override
+        public void getArtifacts(List<String> artifactPaths) {
+            Set<Artifact> artifacts = project.getArtifacts();
+            for (Artifact artifact : artifacts) {
+                try {
+                    artifactPaths.add(artifact.getFile().getCanonicalPath());
+                } catch (IOException e) {
+                    log.error("Unable to resolve project artifact " + e.getMessage());
+                }
+            }
+        }
+
+        @Override
+        public boolean recompileBuildFile(File buildFile, List<String> artifactPaths, ThreadPoolExecutor executor) {
+            try {
+                String modifiedPom = util.readFile(buildFile);
+                XMLUnit.setIgnoreWhitespace(true);
+                XMLUnit.setIgnoreAttributeOrder(true);
+                XMLUnit.setIgnoreComments(true);
+                DetailedDiff diff = new DetailedDiff(XMLUnit.compareXML(this.existingPom, modifiedPom));
+                List<?> allDifferences = diff.getAllDifferences();
+                log.debug("Number of differences in the pom: " + allDifferences.size());
+
+                if (!allDifferences.isEmpty()) {
+                    log.info("Pom has been modified");
+                    MavenProject updatedProject = loadProject(buildFile);
+                    List<Dependency> dependencies = updatedProject.getDependencies();
+                    log.debug("Dependencies size: " + dependencies.size());
+                    log.debug("Existing dependencies size: " + this.existingDependencies.size());
+
+                    List<String> dependencyIds = new ArrayList<String>();
+                    List<Artifact> updatedArtifacts = getNewDependencies(dependencies, this.existingDependencies);
+
+                    if (!updatedArtifacts.isEmpty()) {
+                        for (Artifact artifact : updatedArtifacts) {
+                            if (("esa").equals(artifact.getType())) {
+                                dependencyIds.add(artifact.getArtifactId());
+                            }
+
+                            org.eclipse.aether.artifact.Artifact aetherArtifact = new org.eclipse.aether.artifact.DefaultArtifact(
```

Comment:
When you are resolving dependencies, do you take into account scope (don't add `provided` scope dependencies to the runtime classpath) and non-jar type dependencies (for example, `pom` type dependencies that define transitive dependencies)?

Reply:
We are only dealing with the classpaths for compiling source and tests.  `provided` scope should also be included in this case since they are needed for compiling.  At the moment we do not differentiate between `compile` and `test` scopes (we are using the same set of dependencies for both), but that is something that can be considered in a separate issue.

Transitive dependencies are automatically handled by `repositorySystem.resolveDependencies`